### PR TITLE
Pengram

### PR DESCRIPTION
**Pull request by sjsrey**
A pengram combines [Pen's Parade](https://blogs.worldbank.org/en/developmenttalk/poverty-reduction-growth-and-movements-income-distribution) and a choropleth map.

Functionality
- ascending/descending Pen's Parades
- weighted Pen's Parades
- query functionality for pengram

This is a work in progress for discussion and feedback.

image

### CODE REVIEW

File: inequality/_indices.py
Comment (jGaboardi) [inline inequality/_indices.py:1]: Should we maybe handle these `_indices.py` change over in #69 ?

**Comment by jGaboardi** [comment]
My initial comment is that this is really cool 🎆

File: inequality/pen.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+Serge Rey <srey@sdsu.edu>
+"""
+
+import seaborn as sns
+import matplotlib.pyplot as plt
+import mapclassify as mc
+import matplotlib.patches as patches
+import math
+import pandas as pd
+import numpy as np
```
Comment (martinfleis) [inline inequality/pen.py:23]: We now depend only on lib, numpy and scipy here. All of these will either need to be imported as soft deps in the function or added as required to pyproject. 

The only thing I would be hesitant depending on is seaborn. Do we need it at all or can we use matplotlib/pandas to do the bar plot instead?

**Comment by martinfleis** [comment]
> out of curiosity whats the aversion to seaborn? it's already a dependency in the metapackage because of segregation, and its own footprint is minimal

@knaaptime it is not an aversion to seaborn. That just happened to be a victim of a general rule of trying to avoid additional dependencies to minimise long-term maintenance burden. Stuff can and will break and less stuff we depend on, less issues we will have to solve. I'm likely too conservative in this but I believe that an additional dependency shall be added if we either can't work without it or if working without it would be unnecessarily complicated. I don't think that a single barplot call saving maybe a one line of matplotlib code here is just not enough. It is a similar case we had in mapclassify. In the end, there was no reason for an additional dependency.

File: inequality/pen.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+        ax.set_xticks(first_positions)
+        ax.set_xticklabels(first_labels, rotation=90, fontsize=8)
+
+        ax.set_xlabel(name)
+        ax.set_ylabel(col)
+        ax.set_title(f'Weighted Pen Parade of {name} by {col}')
+
+        # plt.show()
+        return fig
```
[the remark this answers is not in the document]
Comment (sjsrey) [reply inequality/pen.py:168]: I was thinking of that, but wanted to get the inernals right first. If folks thing returning the axes is in line with what is done elsewhere, I can add it.

File: inequality/pen.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    -------
+    matplotlib.axes.Axes
+        A Matplotlib Axes object with the Pen's Parade plot.
+
+    """
+    try:
+        import seaborn as sns
+    except ImportError as e:
+        raise ImportError(
+            "Seaborn is required for pen. Install with  'pip install seaborn'."
```
Comment (jGaboardi) [inline inequality/pen.py:75]:
```suggestion
            "Seaborn is required for pen. Install with 'pip install seaborn'."
```

Also, is there a reason for suggested `pip` over `conda` here and below?

File: inequality/pen.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        ) from e
+
+    try:
+        import matplotlib.pyplot as plt
+    except ImportError as e:
+        error_msg = (
+            "Matplotlib is required for pen. Install with 'pip install " "matplotlib'."
```
Comment (jGaboardi) [inline inequality/pen.py:82]:
```suggestion
            "Matplotlib is required for pen. Install with 'pip install matplotlib'."
```

File: inequality/pen.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        )
+        raise ImportError(error_msg) from e
+
+    try:
+        import pandas as pd
+    except ImportError as e:
+        error_msg = (
+            "Pandas is required for pen. Install it using 'pip install " "pandas'."
```
Comment (jGaboardi) [inline inequality/pen.py:90]:
```suggestion
            "Pandas is required for pen. Install it using 'pip install pandas'."
```

File: inequality/pen.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+        Matplotlib Axes instance to plot on. If None, a new figure and axes
+        will be created. Default is None.
+
+    Returns
+    -------
+    matplotlib.axes.Axes
+        A Matplotlib Axes object with the Pen's Parade plot.
+
+    """
+    try:
```
Comment (jGaboardi) [inline inequality/pen.py:71]: Should this whole `try/except` block be removed and simply call `_check_deps()` as is done in `pentagram()`?

File: pyproject.toml
Comment (jGaboardi) [inline pyproject.toml:1]: Regardless if stick with `seaborn` for this functionality, we at least need to add `matplotlib` and `pandas` as optional deps here. Maybe before `dev` in L59 like:

```
pen = [
    "matplotlib",
    "pandas",
]
```

**Review by jGaboardi — APPROVED** [review]

**Review by knaaptime — APPROVED** [review]

**Pull request merged**